### PR TITLE
[OpenMP][MLIR] Add new arguments to map_info to help support record type maps

### PR DESCRIPTION
**Pull request by agozillon**
This PR adds two new fields to omp.map_info, one BoolAttr and one DenseIntElementsAttr.

The BoolAttr is named partial_map, and is a flag that indicates if the record type captured by
the map_info operation is a partial map, or if it is mapped in its entirety, this currently helps
the later lowering determine the type of map entries that need to be generated.

The DenseIntElementsAttr named members_index is intended to track the placement of each 
member map_info operations (and by extension mapped member variable) placement in the 
parent record type.

### CODE REVIEW

**Comment by agozillon** [comment]
Small ping for attention on this PR please, to start nudging it towards the finish line if there is one in its future. I am sorry ahead of time for stealing some of a reviewers precious time.

**Comment by kparzysz** [comment]
Is the `ptr_and_obj` modifier connected to the new arguments?

**Comment by agozillon** [comment]
> Is the `ptr_and_obj` modifier connected to the new arguments?

Connected to the overall PR, I should alter the description to change that. Or perhaps simplify the stack by moving the changes into a seperate PR (my original thought was that it would add more dependencies to the PR stack, but I'm not necessarily against that if it makes things easier to review). Do let me know which you'd prefer though!

The main goal is to move the setting of the `ptr_and_obj` modifier to the frontend, rather than the backend which would allow us to set it more accurately (it should only be set when necessary but currently we assume all members need the flag in lowering as we do not have the information to indicate if it is necessary, this is only correct for the descriptor base addresses or other pointer members), it's primarily used when mapping declare target/descriptor base addresses currently so unfortunately making this change means some minor alterations to these are mixed into the PR stack (moving the setting of the `ptr_and_obj` bits essentially)

**Comment by kparzysz** [comment]
Yeah, I didn't realize this was a part of a series.  :)

**Comment by agozillon** [comment]
> Yeah, I didn't realize this was a part of a series. :)

Sorry, I should have made it clearer in the title! I was a little worried about title length as I broke the windows clonability of the llvm-project with a combination of bad commit names and SPR with the previous PR stack, so I avoided adding any index prefix like PR 1/4 etc. I think the morale of the story is that I really need to work a lot on my PR/Commit titles moving forward :-)

The full series are the following:


https://github.com/llvm/llvm-project/pull/82853 : the flang level changes required, also the top-level PR that should pass CI if all is well in the stack
https://github.com/llvm/llvm-project/pull/82852 : the MLIR -> LLVM-IR changes required, mainly a refactoring and generalisation of the previous member map work from the initial descriptor type mapping PR
https://github.com/llvm/llvm-project/pull/82851 : the current PR, MLIR changes required
https://github.com/llvm/llvm-project/pull/82850 : runtime tests the patch should enable, or perform more optimally/correctly corresponding to how Clang does things. Will pass CI by default as it doesn't run these tests unfortunately.

**Comment by kparzysz** [comment]
I'm not sure if member placement alone is sufficient.  As I commented in the other PR, for `x%y%z` we need to map the `z` component of the flattened structure of `x`, which we don't have a representation of.

**Comment by agozillon** [comment]
> I'm not sure if member placement alone is sufficient. As I commented in the other PR, for `x%y%z` we need to map the `z` component of the flattened structure of `x`, which we don't have a representation of.

It will be extended in a subsequent PR to have N-D indexing, rather than just the 1-D that's currently here, it's only intended to handle the example cases in https://github.com/llvm/llvm-project/pull/82850 for the moment. Which effectively amounts to only being able to map the first layer of a derived type, so no nested derived type member maps. 

Adding support for derived type member mapping is going to be an iterative process, with support added incrementally, as otherwise the PR would be a little extreme for any reviewer (as well as the fact I am building on the support piece by piece, as has been the case with all of the map work).

Current plan is to add gradual map support for the following cases related to derived types: 
- Nested/recursive derived type member mapping (my bad terminology for the case you brought up)
- descriptor type member mapping (allocatable/pointers/target) (is a case of the above, but will likely come with some caveats due to the nature of these types)
- Recursive derived type mapping of all pointer/allocatable members (whenever a derived type is mapped, all these components are supposed to be mapped as well magically)
- Mixed full struct map with member map (not really something required in Fortran I think, but it's something C/C++ allows in the case where you map a structure then explicitly map data pointed to by a pointer member)

**Comment by agozillon** [comment]
So, incremental increase of support if possible and get feedback as early as possible on direction is the aim :-)

The idea for the member index, at least currently will be to move it to a multi-dimensional index (via DenseIntElementsAttr) as an example:

```
type :: nested
    real(4) :: x1
    real(4) :: y2
    real(4) :: z3
end type nested

type :: array
  real(4) :: x
  real(4) :: y
  real(4) :: z
  type(nested) :: nest 
end type array

```

In the above case array%nest would be index: (3) and array%nest%z3 would be index: (3, 2), if we replaced z3 with another layer through a derived type and mapped a member of it, it would be an index of (3, 2, ...) and so on.

I am hoping that will be reasonable enough to support the case of map(tofrom: x%y%z, x%y) and from a little IR testing it seems it should be, however, I am still working on the initial implementation for this use-case, so I could be wrong. The main thing I am worried about and a little unsure on currently is if the layout of the derived type will always be segmentable like this. But there could be other things wrong with modeling the placement of members like this that I am missing (or a better method of doing so)! So if there is I would love for some feedback.

**Comment by agozillon** [comment]
> So, incremental increase of support if possible and get feedback as early as possible on direction is the aim :-) But if we wanted to avoid that and have it all in a single set of PRs that'd be possible as well, but having this up for an early review is still ideal to see what people feel about the approach.
> 
> The idea for the member index, at least currently will be to move it to a multi-dimensional index (via DenseIntElementsAttr) as an example:
> 
> ```
> type :: nested
>     real(4) :: x1
>     real(4) :: y2
>     real(4) :: z3
> end type nested
> 
> type :: array
>   real(4) :: x
>   real(4) :: y
>   real(4) :: z
>   type(nested) :: nest 
> end type array
> ```
> 
> In the above case array%nest would be index: (3) and array%nest%z3 would be index: (3, 2), if we replaced z3 with another layer through a derived type and mapped a member of it, it would be an index of (3, 2, ...) and so on.
> 
> I am hoping that will be reasonable enough to support the case of map(tofrom: x%y%z, x%y) and from a little IR testing it seems it should be, however, I am still working on the initial implementation for this use-case, so I could be wrong. The main thing I am worried about and a little unsure on currently is if the layout of the derived type will always be segmentable like this. But there could be other things wrong with modeling the placement of members like this that I am missing (or a better method of doing so)! So if there is I would love for some feedback from anyone reading.

Saying the above, I do have an initial implementation (quicker than I anticipated) that improves on this PR to handle this type of mapping and I do hope to open a PR on it in the coming weeks (provided I don't run into any critical failures along the way), it primarily alters the member index generation, member index storage attribute (DenseIntElementsAttr), and the algorithm that works out which member map is first and last of the mapped members.  

I primarily need to test and verify it further and make it presentable by tidying it a little and adding the usual array of tests (the most time consuming component) before it's ready for opening. 

Depending on how far along in the review process this PR is at the time it's ready, I could either update this PR stack with it, wait until this stack has landed to create the PR stack for the next iteration, or add onto the top of this stack with the next one, if any reviewer has any preferences, requests or thoughts on this I would love to hear it. 

I mainly would like to make the review process as easy as possible for you all as I am aware the change set is fairly broad. And to that affect I'll try and seperate out some of the self-reliant but quality of life changes from this stack into seperate PRs if they can be over the next few days, mainly the MAP_AND_PTR_OBJ changes and the bounds refactoring and movement of the createAlteredByCaptureMap function invocation.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (skatrak) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:1044]: Nit: Maybe store a local reference to `membersIdx.getShapedType().getShape()` and reuse it to reduce the verbosity of the code a bit.

**Review by skatrak — APPROVED** [review]
This LGTM, I just have some small nits. Thanks Andrew! Wait for approval by @kparzysz before merging, in case he still has concerns.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (jsjodin) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:1044]: Using names for the local references might help also to make the code clearer.
`i * membersIdx.getShapedType().getShape()[1]` can be hoisted out of the inner loop also.

**Review by jsjodin — APPROVED** [review]

**Review by kparzysz — APPROVED** [review]

**Review by skatrak — APPROVED** [review]
Thank you Andrew for addressing my concerns, LGTM

**Comment by agozillon** [comment]
I made a bit of a mistake not using SPR to land this PR series, so i've ended up having to directly commit the PR, the associated commit for this PR can be found here: https://github.com/llvm/llvm-project/commit/50df0ff515697090983b3cb951c48377bcfb6ddf

**Pull request merged**
